Dynatrace/dynatrace-operator
# Add edgeconnect  provisioner mode

## Description

The operator supports EdgeConnect `Provisioner` mode.  EdgeConnect CR has new flag `spec.oauth.provisioner`. 
If the flag is set to `true` then the operator 
- creates the EdgeConnect entry on the tenant
- updates the EdgeConnect entry on the tenant if `HostPatterns` field is changed in the CR
- removes the EdgeConnect entry on the tenant when the CR is being deleted
- handles the OAuth credentials, provides spec.oauth.clientSecret to the EC deployment in `Regular` mode and an additional secret with credentials from the EC API (CREATE) in `Provisioner` mode

TODO comments point out to locations where error recovery logic should be added (based on `managedByDynatraceOperator` attribute). For example:
- EdgeConnect entry on the tenant created by the operator exists but has to be re-created
- EdgeConnect entry on the tenant having the same name as CR already exist 
- EdgeConnect entry on the tenant has the same name as CR but different ID

OAuth credentials provided by the EC API are not rotated because there is no way to detect the problem.

## How can this be tested?
Make sure your tenant OAuth client has required scopes.
1) Deploy EdgeConnect CR
```
apiVersion: v1
kind: Secret
metadata:
  name: ec-provisioner
  namespace: dynatrace
data:
  oauth-client-id: <client-id>
  oauth-client-secret: <client-secret>
---                                                                                                                                                                                                                                                                                                                                                                          
apiVersion: dynatrace.com/v1alpha1
kind: EdgeConnect
metadata:
  name: ec-provisioner
  namespace: dynatrace
spec:
  apiServer: <tenant>.dev.apps.dynatracelabs.com
  oauth:
    clientSecret: ec-provisioner                                                                                                                                                                                                                                                                                                                                                     
    endpoint: <sso-endpoint>
    resource: <resource>
    provisioner: true
  hostPatterns:
    - "*.internal.org"
```
After a moment 
- the EdgeConnect CR status is `RUNNING`
- `ec-provisioner` deployment is created
- `ec-provisioner-client` secret is created
- `ec-provisioner` entry on the tenant is created
2) Modify CR `hostPatterns` field and apply changes
- `ec-provisioner` entry on the tenant is updated accordingly
3) Delete EdgeConnect CR
- `ec-provisioner` entry on the tenant is also deleted


## Checklist
